core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -255,17 +273,162 @@ private void rebuildIndex() throws Exception {
+      if (ctx.getGlobalConfiguration().serialization().marshaller() == null) {
+         migrateCorruptDataV12_0();
+         return;
+      }
+
+      // Data is not corrupt, so simply update file magic
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startMigratingPersistenceData(cacheName);
+      try {
+         channel.write(ByteBuffer.wrap(MAGIC_LATEST), 0);
+      } catch (IOException e) {
+         throw PERSISTENCE.persistedDataMigrationFailed(cacheName, e);
+      }
+      rebuildIndex();
+   }
+
+   private void migrateCorruptDataV12_0() {
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startRecoveringCorruptPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+
+      // Day before the release of Infinispan 10.0.0.Final
+      // This was the first release that SFS migrations on startup could be migrated from via ISPN 10 -> 11 -> 12
+      long sanityEpoch = LocalDate.of(2019, 10, 26)
+            .atStartOfDay()
+            .atZone(ZoneId.systemDefault())
+            .toInstant()
+            .toEpochMilli();
+      long currentTs = timeService.wallClockTime();
+
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
+         //Write Magic
+         newChannel.truncate(0);
+         newChannel.write(ByteBuffer.wrap(MAGIC_12_0), 0);
+
+         long oldFilePos = MAGIC_12_0.length;
+         while (true) {
+            buf = readChannel(buf, oldFilePos, KEY_POS_LATEST);
+            // EOF reached
+            if (buf.remaining() > 0)
+               break;
+
+            buf.flip();
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            // Semantic check to find valid FileEntry
+            // fe.keyLen = 0 is valid, however it means we should skip the entry
+            if (fe.size <= 0 || fe.keyLen <= 0 || fe.dataLen <= 0 || fe.metadataLen < 0 || fe.internalMetadataLen < 0 || fe.expiryTime < -1) {
+               // Check failed, try to read FileEntry from the next byte
+               oldFilePos++;
+               continue;
+            }
+
+            K key;
+            V value;
+            Metadata metadata = null;
+            ByRef.Long offset = new ByRef.Long(oldFilePos + KEY_POS_LATEST);
+            try {
+               // Read old entry content and then write
+               key = unmarshallObject(buf, offset, fe.keyLen);
+               value = unmarshallObject(buf, offset, fe.dataLen);
+
+               int metaLen = fe.metadataLen > 0 ? fe.metadataLen - TIMESTAMP_BYTES : 0;
+               if (metaLen > 0)
+                  metadata = unmarshallObject(buf, offset, metaLen);
+
+               // Entries successfully unmarshalled so it's safe to increment oldFilePos to FileEntry+offset so brute-force can resume on next iteration
+               oldFilePos = offset.get();
```

Comment:
Technically this line and the same exact one in the if below can be removed and put the exact line after the if. But I am fine leaving it how it is.